src/Compiler/Microsoft.AspNetCore.Razor.Language/src/RazorSourceDocumentProperties.cs

```diff
@@ -54,5 +47,23 @@ public RazorSourceDocumentProperties(string filePath, string relativePath)
     /// The relative path (if provided) is used for display (error messages). The project-relative path may also
     /// be used to embed checksums of the original source documents to support runtime recompilation of Razor code.
     /// </remarks>
-    public string RelativePath { get; }
+    public abstract string? RelativePath { get; }
+
+    private sealed class DefaultRazorSourceDocumentProperties : RazorSourceDocumentProperties
```

Comment:
What's the advantage of doing this vs. having just one sealed `RazorSourceDocumentProperties`?

Reply:
This pattern avoids unnecessary storage when one or both of `FilePath` and `RelativePath` are unspecified. I could certainly see an argument that it's an over-optimization, but I'd prefer to keep it as is unless you and @jaredpar would strongly prefer that we don't do it.

Reply:
Don't feel strongly here either way.

Reply:
Then I'll plan on leaving this as is.